hal/ia32/timer.c

```diff
@@ -368,63 +479,157 @@ time_t hal_timerGetUs(void)
-			_hal_lapicTimerStart(timer.schedulerTimer.lapic.wait[id]);
-			break;
-		default:
-			/* Not supported */
-			break;
+	hal_spinlockSet(&timer_common.sp, &sc);
+	if (timer_common.schedulerTimer->schedulerSetWakeup != NULL) {
+		timer_common.schedulerTimer->schedulerSetWakeup(waitUs);
 	}
-	hal_spinlockClear(&timer.sp, &sc);
+	hal_spinlockClear(&timer_common.sp, &sc);
 }
 
 
+int hal_timerRegister(int (*f)(unsigned int, cpu_context_t *, void *), void *data, intr_handler_t *h)
+{
+	h->f = f;
+	h->n = SYSTICK_IRQ;
+	h->data = data;
+
+	return hal_interruptsSetHandler(h);
+}
+
+
+char *hal_timerFeatures(char *features, unsigned int len)
+{
+	static const char textScheduling[] = "Scheduling timer = ";
+	static const char textTimestamp[] = "\nhal: Timestamp timer = ";
+	const size_t textSchedulingLength = sizeof(textScheduling) - 1;
+	const size_t textTimestampLength = sizeof(textTimestamp) - 1;
+
+	unsigned int off = 0;
+	(void)hal_strncpy(features + off, textScheduling, len);
+	off += textSchedulingLength;
+	len -= textSchedulingLength;
+
+	off += timer_common.schedulerTimer->name(features + off, &len);
+
+	(void)hal_strncpy(features + off, textTimestamp, len);
+	off += textTimestampLength;
+	len -= textTimestampLength;
+
+	off += timer_common.timestampTimer->name(features + off, &len);
+	features[off] = '\0';
+	return features;
+}
+
+
+static const hal_timer_t _hal_pitTimer = {
+	.type = timer_pit,
+	.name = _hal_pitName,
+	.init = _hal_pitInit,
+	.schedulerIrq = _hal_pitTimerIrqHandler,
+	.schedulerSetWakeup = NULL,
+	.schedulerInitCore = NULL,
+	.timestampGetUs = _hal_pitGetUs,
+	.timestampBusyWaitUs = _hal_pitBusyWaitUs,
+};
+
+
+/* Timer used at the very beginning, used only as timestamp for Local APIC Timer */
+static const hal_timer_t _hal_defaultTimer = {
+	.type = timer_undefined,
+	.name = _hal_pitName,
+	.init = _hal_defaultInit,
+	.schedulerIrq = NULL,
+	.schedulerSetWakeup = NULL,
+	.schedulerInitCore = NULL,
+	.timestampGetUs = _hal_pitGetUs,
+	.timestampBusyWaitUs = _hal_pitBusyWaitUs,
+};
+
+
+static const hal_timer_t _hal_lapicTimer = {
+	.type = timer_lapic,
+	.name = _hal_lapicTimerName,
+	.init = _hal_lapicTimerInit,
+	.schedulerIrq = _hal_lapicTimerIrqHandler,
+	.schedulerSetWakeup = _hal_lapicSetWakeup,
+	.schedulerInitCore = _hal_lapicInitCore,
+	.timestampGetUs = _hal_lapicGetUs,
+	.timestampBusyWaitUs = NULL,
+};
+
+
+static const hal_timer_t _hal_hpetTimer = {
+	.type = timer_hpet,
+	.name = _hal_hpetName,
+	.init = _hal_hpetInit,
+	.schedulerIrq = NULL,
+	.schedulerSetWakeup = NULL,
+	.schedulerInitCore = NULL,
+	.timestampGetUs = _hal_hpetGetUs,
+	.timestampBusyWaitUs = _hal_hpetBusyWaitUs,
+};
+
+
+/* List of scheduling timers from the best, to the worst */
+static const hal_timer_t *const _hal_schedulingTimers[] = { &_hal_lapicTimer };
+
+/* List of timestamp timers from the best, to the worst */
+static const hal_timer_t *const _hal_timestampTimers[] = { &_hal_hpetTimer };
```

Comment:
Hmm, I don't think we expect more timers anytime soon, so no need for one-element lists for now